neo/Network/P2P/Payloads/Transaction.cs

```diff
@@ -61,6 +62,37 @@ public UInt256 Hash
             Attributes.GetVarSize() +   //Attributes
             Witnesses.GetVarSize();     //Witnesses
 
+        public void CalculateGas()
+        {
+            if (Sender is null) Sender = UInt160.Zero;
+            if (Attributes is null) Attributes = new TransactionAttribute[0];
+            if (Witnesses is null) Witnesses = new Witness[0];
+            _hash = null;
+            long consumed;
+            using (ApplicationEngine engine = ApplicationEngine.Run(Script, this))
+            {
+                if (engine.State.HasFlag(VMState.FAULT))
+                    throw new InvalidOperationException();
+                consumed = engine.GasConsumed;
+            }
+            _hash = null;
```

Comment:
Is needed duplicate this line?

Reply:
Yes, after setting `Gas`, the hash may be changed, so we set it to `null`.